src/core/bsequence.ml

```diff
@@ -0,0 +1,49 @@
+
+type +'a node =
+    | Nil
+    | Cons of 'a * 'a t
+and 'a t = unit -> 'a node
+
+let empty () = Nil
+
+let return x () = Cons(x, empty)
+
+let rec map f i () = 
+    match i () with 
+    | Nil -> Nil
+    | Cons (x, r) -> Cons (f x, map f r)
+
+let rec filter p i () = 
+    match i () with 
+    | Nil -> Nil
+    | Cons (x, r) -> if p x then Cons (x, filter p r) else filter p r ()
+
+let rec filterMap p i () = 
+    match i () with
+    | Nil -> Nil
+    | Cons (x, r) -> begin
+            match p x with 
```

Comment:
Too much indentation. I also think I'd prefer `begin` on the same line as `match`:
```ml
begin match p x with
| Some c -> ..
| None -> ...
end
```